### ISSUE TITLE
TagPosition not updated properly before bare & character

### ISSUE THREAD

**Issue by jgm** [ndmitchell/tagsoup#66]
The bug is shown by the following example:
```haskell
Prelude Text.HTML.TagSoup> parseTagsOptions  Text.HTML.TagSoup.parseOptions{ optTagWarning = False, optTagPosition = True} "<a>&</a>"
[TagPosition 1 1,TagOpen "a" [],TagPosition 1 5,TagText "&",TagPosition 1 5,TagClose "a"]
```
Note that the TagPosition is the same before and after `TagText "&"`.

I would expect instead:
```haskell
[TagPosition 1 1,TagOpen "a" [],TagPosition 1 4,TagText "&",TagPosition 1 5,TagClose "a"]
```

Compare:
```haskell
Prelude Text.HTML.TagSoup> parseTagsOptions  Text.HTML.TagSoup.parseOptions{ optTagWarning = False, optTagPosition = True} "<a>x</a>"
[TagPosition 1 1,TagOpen "a" [],TagPosition 1 4,TagText "x",TagPosition 1 5,TagClose "a"]
```

This is related to two pandoc bugs, jgm/pandoc#4094, jgm/pandoc#4088.

**Comment by jgm** [comment]
Putting a trace on `parse` in `Specification.hs` gives
```
[Pos (Position 1 1),Tag,Char 'a',Pos (Position 1 3),Pos (Position 1 4),TagEnd,Pos (Position 1 4),Pos (Position 1 5),Char '&',Pos (Position 1 5),TagShut,Char 'a',Pos (Position 1 8),Pos (Position 1 9),TagEnd,Pos (Position 1 9),Char ' ',Pos (Position 1 10)]
```
Note the consecutive `Pos (Position 1 4), Pos (Position 1 5)` which seems wrong.

**Comment by jgm** [comment]
This change seemed to help, but I'm not sure of the other ramifications:
```haskell
 charRef :: Parser -> Bool -> Maybe Char -> S -> [Out]
-charRef resume att end S{..} = pos $ case hd of
+charRef resume att end S{..} = case hd of
     _ | eof || hd `elem` "\t\n\f <&" || maybe False (== hd) end -> '&' & resume s
-    '#' -> charRefNum resume s tl
-    _ -> charRefAlpha resume att s
+    '#' -> pos $ charRefNum resume s tl
+    _ -> pos $ charRefAlpha resume att s
```

**Comment by ndmitchell** [comment]
Reading it, that looks like exactly the right solution, and the test suite doesn't hang for me. I've committed it and a test, to confirm the test suite also runs fine on the CI machines.

**Comment by ndmitchell** [comment]
BTW, thanks for the bug report, excellent analysis, and even the patch :).

**Comment by ndmitchell** [comment]
CI worked flawlessly. Unless you have any reason to the contrary, I plan to make a release with this fix tonight.

**Comment by jgm** [comment]
I had a few other changes for debugging.  When I took those out, the test suite ran fine!

Looks like the same sort of issue affects e.g. the second clause in
```haskell
charRefNum2 resume o hex S{..} = pos $ case hd of
    _ | hexChar hex hd -> (if hex then EntityHex else EntityNum) & hd & charRefNum3 resume hex tl
    _ -> errSeen "&" & '&' & resume o
```
(and similarly in some of the other functions).

```
*Text.HTML.TagSoup> parseTagsOptions parseOptions{ optTagWarning = False, optTagPosition = True} "<a>&#z"
[TagPosition 1 1,TagOpen "a" [],TagPosition 1 6,TagText "&#z"]
```

**Comment by ndmitchell** [comment]
OK, needs some more work here - I've got a list of them now so will have a more complete go tomorrow.

**Comment by ndmitchell** [comment]
Looking harder, I think all the `pos` statements around entity position parsing were wrong, so I removed them all, and it now seems to work as I expect - both in the error cases and the success cases. Let me know what you think.

**Comment by jgm** [comment]
Looks good.  I compiled pandoc against tagsoup master, and
the bugs I had before went away (and no new ones appeared).

Thanks!

+++ Neil Mitchell [Nov 30 17 05:41 ]:

**Comment by ndmitchell** [comment]
Awesome. I've released tagsoup-0.14.2 with those fixes.
